src/Collections/Arrays/BinarySearch.dfy

```diff
@@ -0,0 +1,46 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Wrappers.dfy"
+include "../../Relations/Relations.dfy"
+include "../../Relations/Comparison.dfy"
+
+module Search {
+  export
+    provides BinarySearch, Wrappers
+    provides Wrappers 
+
+    import opened Relations
+    import opened Wrappers
+    import opened Comparison = Relations.Comparison
+
+  method BinarySearch<T>(a: array<T>, key: T, compare: (T, T) -> Comparison.CompResult) returns (r:Option<nat>)
+    requires forall i,j :: 0 <= i < j < a.Length ==> compare(a[i], a[j]).LessThan? || a[i] == a[j]
+    requires Total?(compare)
+    ensures r.Some? ==> r.value < a.Length && a[r.value] == key
+    ensures r.None? ==> key !in a[..]
```

Comment:
Some versions of binary search will return the index where the element WOULD be when it isn't present. If we want to support that here in the future we'd have to change the return type which would be a breaking change, or define a second version. We could use `Result<nat, nat>`, but that definitely wouldn't read as naturally as a more generic `Either<A, B>` type.

Reply:
I've decided I'm fine with having to define a second alternate version if and when we want this functionality.

Reply:
Yes, like `BinarySearchInsertionIndex`, would be good down the road.